### PR TITLE
Fix context.columns typescript issues

### PR DESCRIPTION
**Pull request by nellh**
Implement ColumnsMap class for preserving ordered keys when creating a parseTSV object. ColumnsMap can be constructed to return a proxy which allows the underlying map to be accessed with a property accessor (`map[key]`) like it is a regular object. A limitation of the proxy is that it only allows key access and iteration, other methods on the underlying map are not passed through and some object functions will not work.

### CODE REVIEW

File: bids-validator/src/types/columns.test.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    columns[2] = ['1']
+    columns[1] = ['2']
+    columns['b'] = ['3']
+
+    // Check that the order is still correct with iteration
+    let iteration = 0
+    for (const [_, val] of columns) {
+      assertEquals(val, [iteration.toString()])
+      iteration += 1
+    }
```
Comment (effigies) [inline bids-validator/src/types/columns.test.ts:22]: Do we also have property access, or is it just square-bracket access `[]`?

```suggestion
    }

    // Check property access
    assertEquals(columns.a, '0')
```

Is there a way to get property access, if not?

Comment (nellh) [reply bids-validator/src/types/columns.test.ts:22]: Square bracket and dot access to properties is equivalent in JavaScript but it did point out a bug with how the name is looked up. aba7edc includes a fix and your suggested test change.

**Review by effigies — APPROVED** [review]
I don't fully understand `columns.ts`, but everything else LGTM.

**Comment by effigies** [comment]
Oh, we are getting a lot of:

```
 error: TypeError: context.columns.keys is not a function or its return value is not iterable
  const headers = [...context.columns.keys()]
```

Do we need to find an alternative way to do that, or can we implement `keys`?

**Comment by nellh** [comment]
> Oh, we are getting a lot of:
> 
> ```
>  error: TypeError: context.columns.keys is not a function or its return value is not iterable
>   const headers = [...context.columns.keys()]
> ```
> 
> Do we need to find an alternative way to do that, or can we implement `keys`?

Keys can be passed from the underlying map. Let me fix that...

File: bids-validator/src/schema/applyRules.ts
Comment (rwblair) [inline bids-validator/src/schema/applyRules.ts:1]: Line 200 is failing for me:
`error: Uncaught (in promise) TypeError: context.columns.keys is not a function or its return value is not iterable`

**Comment by nellh** [comment]
Added support for columns.keys() and a test for that but that breaks if you have a "keys" column, fixed it instead by changing to Object.keys(columns).

File: bids-validator/src/types/columns.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    super()
+    const columns = new Map<string, string[]>() as ColumnsMap
+    return new Proxy<ColumnsMap>(columns, columnMapAccessorProxy)
+  }
+}
+
+// Proxy handler to implement ColumnsMapType
+export const columnMapAccessorProxy = {
+  get: function (target: ColumnsMap, prop: any) {
+    if (prop === Symbol.iterator) {
```
Comment (rwblair) [inline bids-validator/src/types/columns.ts]: I had to add a `Reflect.get(target, prop)` and test return of that for typeof function then bind it to get calls to `keys` working. Not sure if there is a more elegant approach.

**Comment by effigies** [comment]
The failing tests are the expected failures at this point.

**Comment by rwblair** [comment]
I can't get object.keys to work:
```
> import { ColumnsMap } from './columns.ts'
undefined
> let x = new ColumnsMap()
undefined
> x.a = 3
3
> Object.keys(x)
[]
```

**Comment by nellh** [comment]
> I can't get object.keys to work:
> 
> ```
> > import { ColumnsMap } from './columns.ts'
> undefined
> > let x = new ColumnsMap()
> undefined
> > x.a = 3
> 3
> > Object.keys(x)
> []
> ```

Fixed that with the right prototype method, we could also just add a function with a less likely to conflict name for this.

**Pull request merged**